File: lib/services/translation_service.dart
import 'package:translator/translator.dart';

class TranslationService {
  static final GoogleTranslator _translator = GoogleTranslator();
  static final Map<String, String> _cache = {};

  // Translate text using Google Translate API
  static Future<String> translate(String text, String targetLanguage) async {
    if (text.isEmpty) return text;
    
    // Check cache first
    final cacheKey = '${text}_$targetLanguage';
    if (_cache.containsKey(cacheKey)) {
      return _cache[cacheKey]!;
    }

    try {
      final translation = await _translator.translate(
        text,
        to: targetLanguage == 'bn' ? 'bn' : 'en',
      );
      
      final translatedText = translation.text;
      _cache[cacheKey] = translatedText;
      return translatedText;
    } catch (e) {
      print('Translation error: $e');
      return text; // Return original text if translation fails
    }
  }

  // Translate multiple texts at once
  static Future<List<String>> translateBatch(
    List<String> texts,
    String targetLanguage,
  ) async {
    final results = <String>[];
    for (final text in texts) {
      final translated = await translate(text, targetLanguage);
      results.add(translated);
    }
    return results;
  }

  // Clear translation cache
  static void clearCache() {
    _cache.clear();
  }
}

File: clean-care-admin/src/services/translationService.ts
import axios from 'axios';

class TranslationService {
    private cache = new Map<string, string>();
    private apiKey = import.meta.env.VITE_GOOGLE_TRANSLATE_API_KEY;

    async translate(text: string, targetLang: 'bn'): Promise<string> {
        if (!text || text.trim() === '') return text;

        const cacheKey = `${text}_${targetLang}`;

        // Check cache first
        if (this.cache.has(cacheKey)) {
            return this.cache.get(cacheKey)!;
        }

        try {
            // If no API key, return original text
            if (!this.apiKey) {
                console.warn('Google Translate API key not configured');
                return text;
            }

            const response = await axios.post(
                `https://translation.googleapis.com/language/translate/v2`,
                {
                    q: text,
                    target: targetLang,
                    format: 'text',
                },
                {
                    params: { key: this.apiKey },
                }
            );

            const translated = response.data.data.translations[0].translatedText;

            // Cache the translation
            this.cache.set(cacheKey, translated);

            return translated;
        } catch (error) {
            console.error('Translation API error:', error);
            // Return original text as fallback
            return text;
        }
    }

    async translateBatch(texts: string[], targetLang: 'bn'): Promise<string[]> {
        if (!this.apiKey) {
            console.warn('Google Translate API key not configured');
            return texts;
        }

        try {
            const response = await axios.post(
                `https://translation.googleapis.com/language/translate/v2`,
                {
                    q: texts,
                    target: targetLang,
                    format: 'text',
                },
                {
                    params: { key: this.apiKey },
                }
            );

            const translations = response.data.data.translations.map(
                (t: any) => t.translatedText
            );

            // Cache all translations
            texts.forEach((text, index) => {
                const cacheKey = `${text}_${targetLang}`;
                this.cache.set(cacheKey, translations[index]);
            });

            return translations;
        } catch (error) {
            console.error('Batch translation error:', error);
            return texts;
        }
    }

    clearCache(): void {
        this.cache.clear();
    }

    getCacheSize(): number {
        return this.cache.size;
    }
}

export const translationService = new TranslationService();
